File: clojure/src/euler/problem_003.clj
(ns euler
  (:use [clojure.contrib.lazy-seqs :only (primes)])
  (:use [clojure.contrib.math :only (sqrt)]))

(defn divisible-by? [number multiple]
  (zero? (rem number multiple)))

(defn prime-factors [n]
  (filter (partial divisible-by? n) (take-while #(< % (sqrt n)) primes)))

(defn largest-prime-factor [n]
  (reduce max (prime-factors n)))

(println (largest-prime-factor 600851475143))

File: ruby/problem_003.rb
require "prime"

module Euler
  refine Numeric do
    def factors
      (1..Math.sqrt(self).to_i).select { |i| self % i == 0 }
    end

    def prime_factors
      factors.select { |i| Prime.prime? i }
    end
  end
end

using Euler

puts 600_851_475_143.prime_factors.max
